### PR TITLE
 Add renderer selection between GeckoView and WebView. Closes #2915

### LINKED ISSUE #2915: Add renderer selection between GeckoView and WebView
### Why/User Benefit/User Problem
As a user, I'd like to receive the latest and greatest web renderer with all new Firefox features.
As a developer or QA engineer, I'd like a means to cleanly select between GeckoView and WebView without messing up app state.

### What / Requirements
There will be a shared preference to select which web view type to use. A developer or QA engineer will have a means to toggle this setting. Also, each build flavor will have a default web view type to use in case the shared preference is unset.

### Acceptance Criteria (how do I know when I’m done?)
When a developer can toggle between GeckoView and WebView, this task will be completed. It is also in scope to restart the app after toggling so the new renderer starts up with a blank, clean slate to avoid switching state-related bugs.

### PR DESCRIPTION
**Pull request by colintheshots**

### CODE REVIEW

File: app/src/gecko/java/org/mozilla/focus/web/Config.kt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.web
+
+object Config {
+    const val defaultNewRenderer = true
```
Comment (sputnikci) [inline app/src/gecko/java/org/mozilla/focus/web/Config.kt]: \[detekt\] INFO: \[style/NamingConventionViolation\] All names in the codebase should be matching the naming convention of the codebase\.

File: app/src/main/java/org/mozilla/focus/gecko/GeckoViewPrompt.java [hunk 2]
```diff
@@ -50,7 +50,7 @@
 import java.util.Date;
 import java.util.Locale;
 
-final class GeckoViewPrompt implements GeckoSession.PromptDelegate {
+public final class GeckoViewPrompt implements GeckoSession.PromptDelegate {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/gecko/GeckoViewPrompt.java]: \[PMD\] INFO: The class 'GeckoViewPrompt' has a Modified Cyclomatic Complexity of 7 \(Highest = 33\)\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/gecko/GeckoViewPrompt.java]: \[PMD\] INFO: The class 'GeckoViewPrompt' has a NCSS line count of 531 \(Highest = 27\)\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/gecko/GeckoViewPrompt.java]: \[PMD\] INFO: The class 'GeckoViewPrompt' has a Standard Cyclomatic Complexity of 7 \(Highest = 33\)\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/gecko/GeckoViewPrompt.java]: \[PMD\] INFO: The class 'GeckoViewPrompt' has a total cyclomatic complexity of 53 \(highest 11\)\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/gecko/GeckoViewPrompt.java]: \[PMD\] INFO: This class has too many methods, consider refactoring it\.

File: app/src/main/java/org/mozilla/focus/web/ClassicWebViewProvider.java [hunk 3]
```diff
@@ -27,17 +29,17 @@
 /**
  * WebViewProvider for creating a WebView based IWebView implementation.
  */
-public class WebViewProvider {
+public class ClassicWebViewProvider implements IWebViewProvider {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/web/ClassicWebViewProvider.java]: \[PMD\] INFO: This class has too many methods, consider refactoring it\.

File: app/src/main/java/org/mozilla/focus/web/ClassicWebViewProvider.java [hunk 4]
```diff
@@ -79,7 +82,7 @@ private static void setupView(WebView webView) {
     }
 
     @SuppressLint("SetJavaScriptEnabled") // We explicitly want to enable JavaScript
-    private static void configureDefaultSettings(Context context, WebSettings settings) {
+    private void configureDefaultSettings(Context context, WebSettings settings) {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/web/ClassicWebViewProvider.java]: \[PMD\] INFO: The method 'configureDefaultSettings\(Context, WebSettings\)' has a NCSS line count of 19\.

File: app/src/main/java/org/mozilla/focus/web/ClassicWebViewProvider.java [hunk 5]
```diff
@@ -124,38 +127,40 @@ private static void configureDefaultSettings(Context context, WebSettings settin
...
+    public void requestMobileSite(Context context, WebSettings settings) {
         settings.setUserAgentString(toggleDesktopUA(settings, false));
         settings.setUseWideViewPort(false);
     }
 
     /**
      * Build the browser specific portion of the UA String, based on the webview's existing UA String.
      */
-    @VisibleForTesting static String getUABrowserString(final String existingUAString, final String focusToken) {
+    public String getUABrowserString(final String existingUAString, final String focusToken) {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/web/ClassicWebViewProvider.java]: \[PMD\] INFO: The method 'getUABrowserString\(String, String\)' has a NCSS line count of 12\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/web/ClassicWebViewProvider.java]: \[PMD\] INFO: The method 'getUABrowserString\(String, String\)' has a cyclomatic complexity of 5\.

File: app/src/main/java/org/mozilla/focus/web/ClassicWebViewProvider.java [hunk 6]
```diff
@@ -189,7 +194,7 @@ public static void requestMobileSite(Context context, WebSettings settings) {
         return TextUtils.join(" ", tokens) + " " + focusToken;
     }
 
-    @VisibleForTesting static String buildUserAgentString(final Context context, final WebSettings settings, final String appName) {
+    private String buildUserAgentString(final Context context, final WebSettings settings, final String appName) {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/web/ClassicWebViewProvider.java]: \[PMD\] INFO: The method 'buildUserAgentString\(Context, WebSettings, String\)' has a NCSS line count of 12\.

File: app/src/main/java/org/mozilla/focus/web/GeckoWebViewProvider.java [hunk 7]
```diff
@@ -43,26 +48,27 @@
 /**
  * WebViewProvider implementation for creating a Gecko based implementation of IWebView.
  */
-public class WebViewProvider {
+public class GeckoWebViewProvider implements IWebViewProvider {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/web/GeckoWebViewProvider.java]: \[PMD\] INFO: The class 'GeckoWebViewProvider' has a NCSS line count of 288 \(Highest = 6\)\.

**Comment by pocmo** [comment]
@colintheshots I think one last missing piece is to update the implementation of `AppConstants.isGeckoBuild()`. That no longer depends on the flavor:
https://github.com/mozilla-mobile/focus-android/blob/master/app/src/main/java/org/mozilla/focus/utils/AppConstants.java#L31-L32

File: .taskcluster.yml [hunk 8]
```diff
@@ -123,7 +123,7 @@ tasks:
       - "secrets:get:project/focus/tokens"
       - "queue:route:index.project.mobile.focus.release.latest"
     payload:
-      maxRunTime: 3600
+      maxRunTime: 7200
```
Comment (pocmo) [inline .taskcluster.yml]: This is the task for releases. You want to update the `maxRunTime` above to make the PR task run longer.

File: app/src/androidTest/java/org/mozilla/focus/helpers/TestHelper.java [hunk 9]
```diff
@@ -156,7 +156,7 @@ public static void waitForWebContent() {
     public static UiObject AddtoHSOKBtn = TestHelper.mDevice.findObject(new UiSelector()
             .resourceId(getAppName() + ":id/addtohomescreen_dialog_add")
             .enabled(true));
-
+
```
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/helpers/TestHelper.java]: \[Checkstyle\] ERROR: Trailing whitespace

**Pull request merged**